### PR TITLE
Add interpreter for ComplexOp.

### LINKED ISSUE #1101: Add  interpreter for ComplexOp
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### LINKED ISSUE #1102: Add  interpreter for Expm1Op
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by karthikrangasai**
Closes #1101 .

We have the following constraints in the spec:

```
(I1) lhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(I2) rhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(C1) lhs and rhs have the same type.
(C2) result and lhs have the same shape.
(C3) the return type should be a complex type of the element type of the lhs i.e. element_type(`result`) = complex_type(element_type(`lhs`)).
```

These constraints are covered by the following tests

```
I1: lhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
I2: rhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
C1: type(lhs) != type(rhs).
C2: shape(result) != shape(lhs).
C3: element_type(`result`) != complex_type(element_type(`lhs`)). (ODS)
```

### CODE REVIEW

File: stablehlo/reference/Element.cpp [hunk 1]
```diff
@@ -591,6 +591,18 @@ Element ceil(const Element &el) {
   return Element(el.getType(), val);
 }
 
+std::complex<double> complex(const Element &e1, const Element &e2) {
```
Comment (ghpvnist) [inline stablehlo/reference/Element.cpp:594]: This function should return an Element object:
```
Element complex(const Element &e1, const Element &e2) {
  auto type = e1.getType();
  auto complexType = ComplexType::get(type);
  if (isSupportedComplexType(complexType))
    return Element(complexType, std::complex<APFloat>(e1.getFloatValue(),
                                                      e2.getFloatValue()));
  report_fatal_error(invalidArgument("Unsupported element type: %s",
                                     debugString(complexType).c_str()));
}
```
Feedback:
* There is no need to check that both `e1Type` and `e2Type` are supported since the ODS (which runs before this file) checks that the operands have the same element (i.e. `SameOperandsElementType`).
* We cannot use [`isSupportedFloatType`](https://github.com/openxla/stablehlo/blob/main/stablehlo/reference/Types.cpp#L46-L49) since this checks for _all_ supported floating-point types (aka f8s, f16, bf16, f32, and f64), but the supported base types for complex types are `f32` and `f64`. Instead we can create a `ComplexType` and check if this type is supported.
* We can use the Element constructor which takes `complex<APFloat>`.

File: stablehlo/tests/ops_stablehlo.mlir [hunk 2]
```diff
@@ -5606,39 +5606,39 @@ func.func @complex(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>) -> tensor
 
 // -----
 
-func.func @complex_int_input(%arg0: tensor<10x10xi32>, %arg1: tensor<10x10xi32>) -> tensor<10x10xcomplex<i32>> {
+func.func @complex_i1_i2(%arg0: tensor<10x10xi32>, %arg1: tensor<10x10xi32>) -> tensor<10x10xcomplex<i32>> {
```
Comment (ghpvnist) [inline stablehlo/tests/ops_stablehlo.mlir:5609]: You may delete all 5 tests here since they are all covered by the ODS following the checklist.

File: stablehlo/tests/interpret_complex.mlir [hunk 3]
```diff
@@ -0,0 +1,18 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
+
+func.func @add_op_test_f64() {
```
Comment (ghpvnist) [inline stablehlo/tests/interpret_complex.mlir:3]: Reusing the example from the spec:
```
func.func @complex_op_test_f64() {
  %lhs = stablehlo.constant dense<[1.0, 3.0]> : tensor<2xf64>
  %rhs = stablehlo.constant dense<[2.0, 4.0]> : tensor<2xf64>
  %result = stablehlo.complex %lhs, %rhs : tensor<2xcomplex<f64>>
  check.expect_almost_eq_const %result, dense<[(1.0, 2.0), (3.0, 4.0)]> : tensor<2xcomplex<f64>>
  func.return
}
```

**Review by ghpvnist — CHANGES_REQUESTED** [review]
@karthikrangasai  Thank you for this PR! This one is a bit more involving than the previous ones. Let me know if you need help understanding anything.

Change the PR description:
From:
```
(I1) lhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(I2) rhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(C1) lhs and rhs have the same type.
(C2) lhs and rhs have the same shape.
(C3) the return type should be a complex type of the element type of the lhs i.e. element_type(`result`) = complex_type(element_type(`lhs`)).
```
To:
```
(I1) lhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(I2) rhs is a tensor of 32 bit floating-point or 64 bit floating-point.
(C1) lhs and rhs have the same type.
(C2) result and lhs have the same shape.
(C3) the return type should be a complex type of the element type of the lhs i.e. element_type(`result`) = complex_type(element_type(`lhs`)).
```

From:
```
I1: lhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
I2: rhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
C1: type(lhs) != type(rhs).
C2: shape(lhs) != shape(rhs).
C3: element_type(`result`) != complex_type(element_type(`lhs`)). (ODS)
```
To:
```
I1: lhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
I2: rhs is not a tensor of 32 bit floating-point or 64 bit floating-point. (ODS)
C1: type(lhs) != type(rhs).
C2: shape(result) != shape(lhs).
C3: element_type(`result`) != complex_type(element_type(`lhs`)). (ODS)
```

Feedback:
* C2 was incorrectly written.

Can you also add a test [here](https://github.com/openxla/stablehlo/blob/main/stablehlo/tests/infer_stablehlo.mlir#L25) below the compare test:

```
// -----

// CHECK-LABEL: @complex
func.func @complex(%arg0: tensor<10x10xf32>, %arg1: tensor<10x10xf32>) -> tensor<10x10xindex> {
  %0 = "stablehlo.complex"(%arg0, %arg1) {} : (tensor<10x10xf32>, tensor<10x10xf32>) -> tensor<10x10xcomplex<f32>>
  // CHECK: types0 = tensor<10x10xcomplex<f32>>
  %1 = "hlo_test_infer.get_return_types"(%0) : (tensor<10x10xcomplex<f32>>) -> tensor<10x10xindex>
  func.return %1 : tensor<10x10xindex>
}
```
Feedback:
* Following the [checklist](https://github.com/openxla/stablehlo/blob/main/docs/reference_checklist.md), adding one positive test if there are none.

**Review by ghpvnist — CHANGES_REQUESTED** [review]

**Review by ghpvnist — APPROVED** [review]

**Review by sdasgup3 — APPROVED** [review]

**Review by ghpvnist — CHANGES_REQUESTED** [review]
Can you rebase this PR? Some changes to look out for: #1440

**Review by ghpvnist — APPROVED** [review]
